### PR TITLE
feat: BulkWriter implementation

### PR DESCRIPTION
**Pull request by craiglabenz**
The core implementation of BulkWriter.

Unlike most Firestore implementation details, this class does not come in an `AsyncBulkWriter` / `BulkWriter` combo-pack. Instead, `BulkWriter` is opaquely async (by creating a ThreadPoolExecutor internally), meaning batches can be committed in parallel without end developers needing to worry about asyncio or event loops.

### CODE REVIEW

**Comment by tseaver** [comment]
@googlebot I consent.

File: google/cloud/firestore_v1/bulk_writer.py [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+        if self._success_callback:
+            for index, result in enumerate(response.write_results):
+                if isinstance(result, WriteResult):
+                    # if `result.update_time` is None, this was probably a DELETE
+                    self._success_callback(batch_references[index], result, self)
+                else:
+                    # failure callback should go here
+                    # when we figure out what the error situation looks like,
+                    # the NO COVER should also be removed.
+                    pass  # pragma: NO COVER
```
Comment (craiglabenz) [inline google/cloud/firestore_v1/bulk_writer.py:128]: Note: Error handling is coming in a subsequent PR.

File: google/cloud/firestore_v1/bulk_writer.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+
+        self._is_sending = False
+        self._executor.shutdown()
+        # Completely release this resource, allowing our sending methods to
+        # easily detect if `flush` has been called and we should re-instantiate
+        # the executor. The reason for this is that `flush` hangs until everything
+        # is sent (and calling `shutdown` is the easiest way to do that), yet,
+        # it should not completely end the life of this BulkWriter. That role
+        # is filled by the `close` method.
+        self._executor = None
```
Comment (craiglabenz) [inline google/cloud/firestore_v1/bulk_writer.py:257]: Note: This method grows moderately more complex after error handling / retries are included, as retries introduce the possibility of more operations being queued potentially long after `flush()` is called and you thought you'd queued the last writes.

File: google/cloud/firestore_v1/bulk_writer.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+        """
+
+        @functools.wraps(fn)
+        def wrapper(self, *args, **kwargs):
+            if self._send_mode == SendMode.parallel:
+                return self._executor.submit(lambda: fn(self, *args, **kwargs))
+            else:
+                return fn(self, *args, **kwargs)
+
+        return wrapper
```
Comment (craiglabenz) [inline google/cloud/firestore_v1/bulk_writer.py:85]: This decorator is only used on one method (`_send_batch()`), but in doing so simplifies multiple call locations.

File: google/cloud/firestore_v1/bulk_writer.py [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+            getattr(self, "_executor", None) or self._instantiate_executor()
+        )
+
+    def _ensure_sending(self):
+        if not self._is_sending:
+            self._ensure_executor()
+            self._send_until_queue_is_empty()
+
+    def _instantiate_executor(self):
+        return concurrent.futures.ThreadPoolExecutor()
```
Comment (craiglabenz) [inline google/cloud/firestore_v1/bulk_writer.py:231]: This serves as a customization hook, should end-users desire to tinker with the parallelization strategy.

**Review by craiglabenz — COMMENTED** [review]
For reviewers, the general code path here is as follows:

1. Users call [`my_bulk_writer.create(...)`](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L363-L365) (or any of the other CRUD methods)
2. The `BulkWriter` instance [adds this new write operation to an in-progress `BulkWriteBatch` instance](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L388), and finalizes that batch in one of two scenarios:
  a. [There are 20 such pooled write operations](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L267-L277), or
  b. The new write operation [touches a document already involved in a different write operation currently queued](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L385-L386)
3. Once a `BulkWriteBatch` instance is created, it is added to a queue ([`my_bulk_writer._queued_batches`](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L285)), and the first domino in the sending chain is called, [`_send_until_queue_is_empty()`](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L225-L228).
4. This method, `_send_until_queue_is_empty()`, [calls `_request_send()`](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L326), which blocks until all rate-limiting rules agree the batch is able to be sent. Note that we are not yet parallelized, so this properly blocks.
5. The first domino method, `_send_until_queue_is_empty()`, [calls the first method that is parallelized, `_send_ batch()`](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L330). A few notes about this new method:
  a. `_send_batch()` is [optionally (and automatically!) parallelized with a decorator](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L87) that [re-routes execution to `my_bulk_writer._executor`](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L80-L83).
  b. `_send_batch()` routes the RPC's response to [`_process_response()`](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L105), which begins [invoking registered callbacks](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L117-L128) for each write in the batch. All of this work still happens in a different thread via the executor, so the main thread is not held up by any of this work.
6. Now back in `_send_until_queue_is_empty()` and thus, on the main thread, the `BulkWriter` instance [checks to see if the queue is in fact empty, and, if it is not, calls itself](https://github.com/googleapis/python-firestore/blob/325f1992ebf9fdedcd80ec53db04fb20a6997494/google/cloud/firestore_v1/bulk_writer.py#L332-L336).

These are the core ideas behind the current implementation, but a few more moving parts are added in the next PR which allow for error handling and retries.

**Review by crwilcox — APPROVED** [review]

**Comment by crwilcox** [comment]
Reviewed over GVC. Thanks @craiglabenz

**Pull request merged**